File: 394-Decode-String/decodeString.py
class Solution:
    def decodeString(self, s: str) -> str:
        curr_string = ''
        curr_num = 0
        stack = []
        
        for char in s:
            if char.isdigit():
                curr_num = curr_num * 10 + int(char)
            elif char == '[':
                stack.append(curr_num)
                stack.append(curr_string)
                curr_string = ''
                curr_num = 0
            elif char == ']':
                prev_string = stack.pop()
                prev_num = stack.pop()
                curr_string = prev_string + curr_string * prev_num
            else:
                curr_string += char
    
        return curr_string

File: 394-Decode-String/decodeString.js
var decodeString = function(s) {
  let curr_num = 0;
  let curr_string = '';
  let stack = [];
  
  for (let i = 0; i < s.length; i++) {
      let ch = s.charAt(i);
      if (ch >= '0' && ch <= '9') {
          curr_num = curr_num * 10 + parseInt(ch);
      } else if (ch === '[') {
          stack.push(curr_string);
          stack.push(curr_num);
          curr_num = 0;
          curr_string = '';
      } else if (ch === ']') {
          prev_num = stack.pop();
          prev_string = stack.pop();
          curr_string = prev_string + curr_string.repeat(prev_num);
      } else {
          curr_string += ch;
      }
  }
  
  return curr_string;
};